From 77c82e227d56000c4b946e60f62f84053a4b313d Mon Sep 17 00:00:00 2001
From: kkrishnakumar <kkrishnakumar@maxlinear.com>
Date: Wed, 7 Aug 2024 12:27:25 +0530
Subject: [PATCH] WLANRTSYS-81551 Fix to handle channel switch to DFS channels
 when mlo is enabled.

Issue : Unable to switch to DFS channels when mlo is configured.

Root cause : Whenever we try to move to DFS channels, both the main and sibling vap gets removed, and upon adding, only the main vap is added which in turn leads to an issue where MLD is not formed as sibling has not been added.

Fix : To make use of ZWDFS antenna to perform background CAC and channel change to DFS channels whenever MLO is configured so that we can avoid the above scenario. Also enabled background CAC by default in WAV700

FIXUP to WLANRTSYS-63318 ML CSA with puncturing and BSS critical update
---
 hostapd/ctrl_iface.c | 14 ++++++++++++++
 src/ap/hostapd.c     |  8 +++++++-
 2 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index ef02b67a1..73493edc4 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -3072,6 +3072,20 @@ static int hostapd_ctrl_iface_chan_switch(struct hostapd_iface *iface,
 
 	iface->block_tx = false;
 	if (dfs_range) {
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+		struct hostapd_data *hapd = NULL;
+		int j, mlo_enable = 0;
+		for (j = 0; j < iface->num_bss; j++) {
+			hapd = iface->bss[j];
+			if (hapd->conf->mxl_bss_conf.mlo_enable) {
+				mlo_enable = 1;
+				break;
+			}
+		}
+		if (mlo_enable) {
+			switch_type_zwdfs = true;
+		}
+#endif /* CONFIG_IEEE80211BE_MXL_MLO */
 		ret = ieee80211_freq_to_chan(settings.freq_params.freq, &chan);
 		if (ret == NUM_HOSTAPD_MODES) {
 			wpa_printf(MSG_ERROR,
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 8db4d83ee..1cb735fba 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2519,7 +2519,13 @@ static int setup_interface(struct hostapd_iface *iface)
 #ifdef CONFIG_VENDOR_MXL
 	if (mxl_hostapd_setup_interface(hapd))
 		return -1;
-#endif
+
+	/* Enable background CAC by default for WAV700 */
+	if (iface->drv_flags2 & WPA_DRIVER_FLAGS2_RADAR_BACKGROUND) {
+		wpa_printf(MSG_DEBUG, "Enabling background_CAC\n");
+		iface->conf->mxl_conf.background_cac = 1;
+	}
+#endif /* CONFIG_VENDOR_MXL */
 
 	if (hapd->iconf->country[0] && hapd->iconf->country[1]) {
 		char country[4], previous_country[4];
-- 
2.43.0

